File: validation/equilibrium_cases.py
"""Generate independent Cantera references for native ideal-gas equilibrium.

Run with the same mechanism files used by Julia, e.g.:
    python validation/equilibrium_cases.py mechanism output_directory

The mixture cases reproduce the calculations in Cantera's thermo/
equivalenceRatio.py. Equilibrium modes follow ThermoPhase.equilibrate:
https://cantera.org/stable/python/thermo.html#cantera.ThermoPhase.equilibrate
https://cantera.org/stable/examples/python/thermo/equivalenceRatio.html
No reference compositions are used to initialize the native equilibrium solver.
"""
from pathlib import Path
import sys
import cantera as ct
import numpy as np

mechanisms, output = map(Path, sys.argv[1:3])
output.mkdir(parents=True, exist_ok=True)
MODES = ("TP", "TV", "HP", "UV", "SP", "SV")


def equilibrium_cases(name):
    gas = ct.Solution(str(mechanisms / f"{name}.yaml"))
    cases = []
    if name == "h2o2":
        for composition in ("H2:2,O2:1", "H2:1.1,O2:1,AR:5"):
            for mode in MODES:
                # Entropy-conserving reactions can cool below the NASA fit range
                # at room temperature, so use a hotter inlet for SP/SV.
                cases.append((mode, 1200.0 if mode in ("SP", "SV") else 300.0,
                              ct.one_atm, composition))
        for composition in ("H2:1", "O2:1", "AR:1", "H2:1,O2:1e-14"):
            cases.append(("TP", 1000.0, ct.one_atm, composition))
        for temperature in (300., 800., 1500., 2500., 3500.):
            for pressure in (.01, 30.):
                cases.append(("TP", temperature, pressure*ct.one_atm,
                              "H2:2,O2:1,AR:3.76"))
    else:
        for phi in (.5, 1., 2.):
            gas.set_equivalence_ratio(phi, "CH4", "O2:1,N2:3.76")
            composition = gas.X.copy()
            for mode in MODES:
                cases.append((mode, 1500.0 if mode in ("SP", "SV") else 300.0,
                              5*ct.one_atm, composition))
        cases.extend(("TP", t, p*ct.one_atm, "CO:1,CO2:1,N2:2")
                     for t, p in ((300., 1.), (1200., .1), (2500., 30.)))
        cases.append(("TP", 1200., ct.one_atm, "CH4:1,O2:2,N2:7.52,AR:1e-16"))

    data = {key: [] for key in ("mode", "T0", "P0", "X0", "T", "P", "X", "Y", "h", "u", "s", "rho")}
    for mode, temperature, pressure, composition in cases:
        gas.TPX = temperature, pressure, composition
        data["mode"].append(MODES.index(mode)+1)
        data["T0"].append(gas.T)
        data["P0"].append(gas.P)
        data["X0"].append(gas.X.copy())
        gas.equilibrate(mode, rtol=1e-11, max_steps=2000)
        for key, value in (("T", gas.T), ("P", gas.P), ("X", gas.X), ("Y", gas.Y),
                           ("h", gas.enthalpy_mass), ("u", gas.int_energy_mass),
                           ("s", gas.entropy_mass), ("rho", gas.density)):
            data[key].append(value.copy() if isinstance(value, np.ndarray) else value)
    arrays = {key: np.stack(value, axis=1) if key in ("X0", "X", "Y") else np.array(value)
              for key, value in data.items()}
    arrays["cantera_version_utf8"] = np.frombuffer(ct.__version__.encode(), dtype=np.uint8)
    np.savez(output / f"equilibrium-{name}.npz", **arrays)
    print(f"{name}: {len(cases)} equilibrium references ({ct.__version__})", flush=True)


for mechanism in ("h2o2", "gri30"):
    equilibrium_cases(mechanism)

gas = ct.Solution(str(mechanisms / "gri30.yaml"))
mixtures = []
for phi, fuel, oxidizer, kwargs in (
    (1., "CH4", "O2:0.21,N2:0.79", {}),
    (1., "CH4", "O2:0.233,N2:0.767", {"basis": "mass"}),
    (2.5, "CH4:1,O2:0.01,CO:0.05,N2:0.1", "O2:0.2,N2:0.8,CO2:0.05,CH4:0.01", {}),
    (2., "H2", "O2", {"diluent": "H2O", "fraction": {"diluent": .3}}),
    (2., "H2", "O2", {"diluent": "CO2:.5,H2O:.5", "fraction": {"fuel": .1}, "basis": "mass"}),
    (2., "H2:.5,H2O:.5", "O2:.21,N2:.79", {}),
):
    gas.set_equivalence_ratio(phi, fuel, oxidizer, **kwargs)
    mixtures.append(gas.X.copy())
gas.set_mixture_fraction(.055, "CH4", "O2:.21,N2:.79")
mixtures.append(gas.X.copy())
np.savez(output / "composition-gri30.npz", X=np.stack(mixtures, axis=1))
print("gri30: 7 composition references", flush=True)

File: validation/equilibrium_cases.jl
using Arrhenius, NPZ, LinearAlgebra, Test
BLAS.set_num_threads(1)
mechanisms, references = ARGS[1:2]
modes = (:TP,:TV,:HP,:UV,:SP,:SV)

@testset "Cantera ideal-gas equilibrium reference" begin
    for name in ("h2o2","gri30")
        gas = CreateSolution(joinpath(mechanisms,name*".yaml"))
        data = npzread(joinpath(references,"equilibrium-"*name*".npz"))
        version = String(vec(UInt8.(data["cantera_version_utf8"])))
        println("Checking ",name," against Cantera ",version)
        for i in eachindex(data["T0"])
            mode = modes[Int(data["mode"][i])]
            @testset "$name/$i/$mode" begin
                T0, P0, X0 = data["T0"][i], data["P0"][i], data["X0"][:,i]
                result = equilibrate(gas; T=T0,P=P0,X=X0,mode)
                @test result.T ≈ data["T"][i] rtol=2e-7 atol=2e-5
                @test result.P ≈ data["P"][i] rtol=2e-7
                @test maximum(abs.(result.X-data["X"][:,i])) < 1e-7
                @test maximum(abs.(result.Y-data["Y"][:,i])) < 1e-7
                for (key, property) in (("h",cal_hmass_mean),("u",cal_umass_mean),("s",cal_smass_mean))
                    @test property(gas,result.T,result.P,result.X) ≈ data[key][i] rtol=2e-7 atol=2e-2
                end
                rho = result.P*dot(gas.MW,result.X)/(R*result.T)
                @test rho ≈ data["rho"][i] rtol=2e-7
                initial_elements = gas.ele_matrix*X0/dot(gas.MW,X0)
                final_elements = gas.ele_matrix*result.X/dot(gas.MW,result.X)
                @test final_elements ≈ initial_elements rtol=2e-9 atol=1e-14
                for e in eachindex(initial_elements)
                    if initial_elements[e] > 0
                        @test final_elements[e] ≈ initial_elements[e] rtol=2e-8 atol=0
                    else
                        @test final_elements[e] == 0
                    end
                end
                @test all(>=(0),result.X)
                if mode in (:TV,:UV,:SV)
                    @test rho ≈ P0*dot(gas.MW,X0)/(R*T0) rtol=1e-10
                end
                # TP/TV equilibria must minimize the appropriate free energy.
                property = mode == :TP ? cal_gmass_mean : cal_amass_mean
                if mode in (:TP,:TV)
                    @test property(gas,result.T,result.P,result.X) <= property(gas,T0,P0,X0)+1e-5
                end
            end
        end
    end
end

@testset "Cantera equivalence-ratio example" begin
    gas = CreateSolution(joinpath(mechanisms,"gri30.yaml"))
    data = npzread(joinpath(references,"composition-gri30.npz"))["X"]
    cases = (
        (1.,"CH4","O2:.21,N2:.79",(;)),
        (1.,"CH4","O2:.233,N2:.767",(;basis=:mass)),
        (2.5,"CH4:1,O2:.01,CO:.05,N2:.1","O2:.2,N2:.8,CO2:.05,CH4:.01",(;)),
        (2.,"H2","O2",(;diluent="H2O",fraction=(diluent=.3,))),
        (2.,"H2","O2",(;diluent="CO2:.5,H2O:.5",fraction=(fuel=.1,),basis=:mass)),
        (2.,"H2:.5,H2O:.5","O2:.21,N2:.79",(;)),
    )
    for (i,(phi,fuel,oxidizer,kwargs)) in enumerate(cases)
        X = set_equivalence_ratio(gas,phi; fuel,oxidizer,kwargs...)
        @test X ≈ data[:,i] rtol=1e-12 atol=1e-14
        if i in (4,5)
            @test equivalence_ratio(gas,X;fuel,oxidizer,include_species=["H2","O2"]) ≈ phi
        else
            basis = get(kwargs,:basis,:mole)
            @test equivalence_ratio(gas,X;fuel,oxidizer,basis) ≈ phi
            eq = equilibrate(gas;T=300.,P=one_atm,X,mode=:HP)
            @test equivalence_ratio(gas,eq.X;fuel,oxidizer,basis) ≈ phi rtol=1e-8
        end
    end
    X = set_mixture_fraction(gas,.055;fuel="CH4",oxidizer="O2:.21,N2:.79")
    @test X ≈ data[:,7] rtol=1e-12 atol=1e-14
    @test mixture_fraction(gas,X;fuel="CH4",oxidizer="O2:.21,N2:.79") ≈ .055
    @test mixture_fraction(gas,X;fuel="CH4",oxidizer="O2:.21,N2:.79",element="C") ≈ .055
end
